### PR TITLE
`QueryControls`: refactor away from `lodash.groupBy`

### PR DESCRIPTION
**Pull request by flootr**
## What?

This PR removes Lodash's `_.groupBy()` from the `buildTermsTree` function used in `QueryControls`

## Why?

Lodash is known to unnecessarily inflate the bundle size of packages, and in most cases, it can be replaced with native language functionality. See these for more information and rationale:

* https://github.com/WordPress/gutenberg/issues/16938#issuecomment-602837246
* https://github.com/WordPress/gutenberg/issues/17025
* https://github.com/WordPress/gutenberg/issues/39495 

## How?

Replaces `_.groupBy` with a native `.reduce` function.

## Testing Instructions

1. Verify unit tests still pass
2. Verify stories for `QueryControls` still work as expected

### CODE REVIEW

File: packages/components/src/query-controls/terms.ts
Comment (tyxla) [inline packages/components/src/query-controls/terms.ts]: I wonder if we could just inline `ensureParents()`. For me it was easier to read this line:

```
terms.every( ( term ) => term.parent !== null )
```

than having to read what `ensureParents()` does. It's not immediately clear what the function does too - does it add a parent to each term? Or does it filter only terms with parents? I wouldn't expect from a function with this name to return `true` only when ALL passed terms have parents. Inlining the condition will solve this mystery IMHO.

File: packages/components/src/query-controls/terms.ts
Comment (tyxla) [inline packages/components/src/query-controls/terms.ts:30]: The shorter version wins 👍

File: packages/components/src/query-controls/terms.ts
Comment (tyxla) [inline packages/components/src/query-controls/terms.ts:4]: I'd suggest adding a CHANGELOG entry.

**Review by tyxla — APPROVED** [review]
Looks good to me 👍 

There's a couple of suggestions to handle before merging this, but I think it looks great and ready to ship afterwards 🚀

File: packages/components/src/query-controls/terms.ts
Comment (flootr) [reply packages/components/src/query-controls/terms.ts]: Inlining didn't work for me because TypeScript doesn't infer that `parent` isn't `null` anymore. That's why I created a type guard. That's the important part:

> `terms is ( TermWithParentAndChildren & { parent: number } )[]` (L13)

File: packages/components/src/query-controls/terms.ts
Comment (flootr) [reply packages/components/src/query-controls/terms.ts:4]: Agreed 👍🏻 

Update: https://github.com/WordPress/gutenberg/pull/48779/commits/9f9e3c83df58cbfb1e289759e47ac59cbd64742a

File: packages/components/src/query-controls/terms.ts
Comment (tyxla) [reply packages/components/src/query-controls/terms.ts]: I see. Let's think of a name that matches the purpose better then.

Comment (ciampo) [reply packages/components/src/query-controls/terms.ts]: I have an idea — what if we tweaked the way the `termsByParent` dictionary is created, by literally using the `'null'` string as a fallback key when the `parent` is actually `null` ?

<details>
<summary>Something like this</summary>

```diff
diff --git a/packages/components/src/query-controls/terms.ts b/packages/components/src/query-controls/terms.ts
index 951ca626be..9f5688fc91 100644
--- a/packages/components/src/query-controls/terms.ts
+++ b/packages/components/src/query-controls/terms.ts
@@ -8,11 +8,6 @@ import type {
 	TermsByParent,
 } from './types';
 
-const ensureParents = (
-	terms: TermWithParentAndChildren[]
-): terms is ( TermWithParentAndChildren & { parent: number } )[] => {
-	return terms.every( ( term ) => term.parent !== null );
-};
 /**
  * Returns terms in a tree form.
  *
@@ -29,22 +24,23 @@ export function buildTermsTree( flatTerms: readonly ( Author | Category )[] ) {
 			id: String( term.id ),
 		} ) );
 
-	if ( ! ensureParents( flatTermsWithParentAndChildren ) ) {
-		return flatTermsWithParentAndChildren;
-	}
-
 	const termsByParent = flatTermsWithParentAndChildren.reduce(
 		( acc: TermsByParent, term ) => {
 			const { parent } = term;
-			if ( ! acc[ parent ] ) {
-				acc[ parent ] = [];
+			const parentKey = parent ?? 'null';
+			if ( ! acc[ parentKey ] ) {
+				acc[ parentKey ] = [];
 			}
-			acc[ parent ].push( term );
+			acc[ parentKey ].push( term );
 			return acc;
 		},
 		{}
 	);
 
+	if ( termsByParent.null && termsByParent.null.length ) {
+		return flatTermsWithParentAndChildren;
+	}
+
 	const fillWithChildren = (
 		terms: TermWithParentAndChildren[]
 	): TermWithParentAndChildren[] => {

```
</details>

We could even decide to use another string instead of `'null'` — there shouldn't be any risk of conflict with other dictionary keys, since`parent` should be a `number`

Comment (flootr) [reply packages/components/src/query-controls/terms.ts]: I think that's essentially how it was solved before. I'll have another look.

Comment (ciampo) [reply packages/components/src/query-controls/terms.ts]: > I think that's essentially how it was solved before.

Yup, it looks like it. We could keep the same logic, but make it more explicit with a better key name than `'null'` ? It could be a quick way to refactor this PR while introducing the least amount of runtime changes possible.

Comment (flootr) [reply packages/components/src/query-controls/terms.ts]: I do think that's a good use case for `.every` (or `.some`) and drop out early before constructing the object and then throwing it away (in case). I'll have another look and see what I come up with.

Comment (ciampo) [reply packages/components/src/query-controls/terms.ts]: Sounds good. It looks like the `buildTermsTree` already has some unit tests, so we should have good confidence in applying the changes you suggested!

Comment (flootr) [reply packages/components/src/query-controls/terms.ts]: In https://github.com/WordPress/gutenberg/pull/48779/commits/a3745b0f8ce3d082cb511bfd832676110518f480 I tried to find a better name for the type guard and added a comment about why we're using it. I think that should address the original concern. Please let me know if there's any other feedback!

File: packages/components/CHANGELOG.md
Comment (tyxla) [inline packages/components/CHANGELOG.md]: Nit:
```suggestion
-   `QueryControls`: Refactor away from Lodash (`.groupBy`) ([#48779](https://github.com/WordPress/gutenberg/pull/48779)).
```

File: packages/components/src/query-controls/terms.ts
Comment (tyxla) [reply packages/components/src/query-controls/terms.ts]: I'd name the guard function `areAllParentsDefined` or `allTermsHaveParents` but then I'd leave it to your judgment.

**Review by tyxla — APPROVED** [review]
Looks good from my perspective 👍 

cc @ciampo for a final review since he also had some prior feedback.

**Review by ciampo — APPROVED** [review]
> cc @ciampo for a final review since he also had some prior feedback.

🚀 

Thank you for the extra-effort in improving the readability of the code!

**Comment by tyxla** [comment]
I've just opened a very similar PR in the editor package: #49224.

Those should likely come from the same place, but it was like that before, so this is definitely work for another PR.

**Pull request merged**
